Support for forward and backward compatibility in DCL schemes (#540)
zigbee-alliance/distributed-compliance-ledger: docs/design/schema-compatibility.md

# Support for forward and backward compatibility in DCL schemes

Schema changes can cover a wide range of modifications with varying impacts on application compatibility and data integrity. Below are use cases with strategies to manage schema changes and ensure compatibility.

> ## Multiple versions can live in parallel

Reviewer: It would be great to add number for every option.
Like

I. Multiple versions can live in parallel
  1. Strategy for Compatible Changes
     - Option A
     - Option B
  2.  Strategy for Non-Compatible Changes
       - Option C
       - Option D

II. New version replaces the legacy one (V2 replaces V1)
   1. Strategy for Compatible or Convertable changes
        - Option E: Not keeping backward compatibility in API
        - Option F: Keeping backward compatibility in API
   2. Strategy for Non-convertable changes
        - Option G: Not keeping backward compatibility in API
        - Option H: Keeping backward compatibility in API

For changes that are backward-compatible, such as adding optional fields or extending enumerations.

**Description:**

This strategy focuses on updating the schema without ensuring backward compatibility at the API level. Since the schemas are compatible, there will likely be no need for migration.

**Strategy steps:**

- For each update:
  - Update the schema by introducing compatible changes (such as adding a new optional field).
  - Migrate old states to the newer if needed.
  - Update transactions and queries if needed.

### Strategy for Non-Compatible changes (Not keeping backward compatibility in API)

For changes that affect compatibility, like adding mandatory fields.

**Description:**

This strategy focuses on updating the schema without ensuring backward compatibility at the API level. Since the schemas are not compatible, migration is carried out manually through a special transaction.

**Strategy steps:**

- For each update:
  - Update the schema by introducing changes.
  - Update transactions and queries if needed.
  - Add a new transaction to fulfill new required fields (essentially this is a manual migration via transactions)

> ### Strategy for Non-Compatible changes (Keeping backward compatibility in API)

Reviewer: It seems this strategy is suitable for compatible (actually convertable) changes only.
So, probably we can rename the Options here a bit:

- Strategy for Compatible or Convertable changes
  - Option 1: Not keeping backward compatibility in API
  - Option 2: Keeping backward compatibility in API
- Strategy for Non-convertable changes
  - Option 1: Not keeping backward compatibility in API
  - Option 2: Keeping backward compatibility in API

I think we can add the following for the last option:
It's not possible to replace an old version here. "Multiple versions can live in parallel: Strategy for Non-Compatible Changes" options should be used instead.

# Support for forward and backward compatibility in DCL schemes

Schema changes can cover a wide range of modifications with varying impacts on application compatibility and data integrity. Below are use cases with strategies to manage schema changes and ensure compatibility.

## Strategy for Compatible Changes

For changes that are backward-compatible, such as adding optional fields or extending enumerations:

**Strategy steps:**

- One time actions:

>   - Add a version field to DCL schema to track the schema version.

Reviewer: ```suggestion
  - Add an optional version field to all DCL schema to track the schema version.
```

Author: Done

- For each update:

>   - Update the schema.

Reviewer: ```suggestion
  - Update the schema by introducing compatible changes (such as adding a new optional field)
```

Author: Done

>   - Probably need to update transactions and queries.

Reviewer: Let's also add the following:
- DCL doesn't fulfill the Schema version automatically
- It will be up to the transaction submitter (Vendor) to specify a correct Schema version
- If Schema Version is not set - then the initial version (version 0 or 1) is assumed
- It will be up to the client application to process the Schema version

Author: Done

This strategy is straightforward and quick to implement, but only suitable for compatible changes.

## Strategy for Convertible Changes

For changes that affect compatibility but can be converted, like field type changes, renaming fields, and changing enumerations:

**Strategy steps:**

- For each update:

>   - Create a new schema version and state

Reviewer: ```suggestion
  - Create a new version of a Schema and state (a new .proto file)
```

Author: Done

>   - Migrate older states to newer schema version.

Reviewer: Probably we need a clarification here.
Does it mean that a data will be represented in two copied: old format AND new format? So it can be queried in either old (legacy) or new way?

>   - Implement transactions and queries capable of converting data between the latest and older schema version.

Reviewer: If we do the migration of old data to the new format, then probably no need to do migration "on the fly" in handlers?
Also, please note that migration on the fly won't allow to use a Light Client feature (as the data is not in the State and proofs can be generated).

> 

Reviewer: Please clarify, that there will be separate calls/requests for each version, for example
- models/vid/pid
- modelsV2/vid/pid
- modelsV3/vid/pid

As the data is convertible and migrated, old (legacy) entities can be returned in either legacy format (for example `models/vid/pid` or `modelsV2/vid/pid`), or a new format (for example `modelsV3/vid/pid`).

Author: Done

The specific implementation details will vary with each change, each time requires an individual approach.

## Strategy for Handling Non-Convertible Changes

For significant changes that directly impact compatibility, such as adding mandatory fields or removing fields:

**Strategy steps:**

- One time actions:
  - Create a more flexible, generic schema structure to hold a wide range of data formats (Can be used [Any](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/any.proto) as described in [ADR-19](https://docs.cosmos.network/v0.47/build/architecture/adr-019-protobuf-state-encoding#usage-of-any-to-encode-interfaces))
  - Migrate old states to the newer, generic schema.
  - Implement transactions and queries that can handle data according to its version, including mechanisms for converting generic values into the corresponding schema version.
- For each update:
  - Create a new schema version

>   - Probably need to update transactions and queries

Reviewer: Please mention cons here, such as that there will be no strict DCL schema anymore.
Also, this change is a breaking change on its own (for all existing clients).
I would keep this option on the doc, but would move it to "Not Recommended for Production" section.

While offering a robust solution for handling radical changes, this method requires careful planning and development, which can potentially take a significant amount of time.

## Strategy for Structural Changes

For major organizational changes to the schema, such as splitting or merging schemas:

**Strategy steps:**

- For each update:

>   - Create a new schema version and state

Reviewer: ```suggestion
  - Create a new schema version and state (new .proto files)
```

Author: Done

>   - Migrate older states to newer schema version.

Reviewer: Do we really need migration here? It seems we can not always do a migration if for example a new mandatory field is added.

>   - Implement transactions and queries specifically designed for the reorganized schema.

Reviewer: Please clarify, that there will be separate calls for each version, for example
- models/vid/pid
- modelsV2/vid/pid
- modelsV3/vid/pid

As the data is not convertible and not migrated, every entity will be associated with one version only. So, it will be returned by either `models/vid/pid`, `modelsV2/vid/pid` or `modelsV3/vid/pid` depending on what Schema version (.proto version) was used during creation.

Author: Done

These types of changes are relatively rare. Implementing backward compatibility and migration can be complex.

## Conclusion

> To lay the foundation for future compatibility improvements, it's a good idea to start by adding a version field to each schema. This step not only provides compatible updates to schemas, but also prepares the schemas for seamless integration with upcoming compatibility enhancements.

Reviewer: Please also add the following:
Non-compatible changes will be handled by creating new Schemas (new .proto files) with separate queries/requests.

  - It will be up to the transaction submitter (Vendor) to specify a correct Schema version
  - If Schema Version is not set - then the initial version (version 0 or 1) is assumed
  - It will be up to the client application to process the Schema version

This strategy is straightforward and quick to implement, but only suitable for compatible changes.

### Strategy for Convertible Changes

For changes that affect compatibility but can be converted, like renaming fields, and changing enumerations:

**Strategy steps:**

- For each update:
  - Create a new version of a Schema and state (a new .proto file)
  - Migrate older states to newer schema version.
  - Remove the states associated with the older schema versions.
  - Implement transactions and queries for the new schema version.
  - Update older transactions and queries to converting data between the latest and older schema version,  ensuring backward compatibility.
  - There will be separated API for each version of the schema, for example::
    - models/vid/pid
    - modelsV2/vid/pid
    - modelsV3/vid/pid

The specific implementation details will vary with each change, each time requires an individual approach.

> Support for the Light Client feature will not be extended to legacy APIs due to on-the-fly data migration (as the data is not in the State and proofs can be generated).

Reviewer: Not sure that it's a good option....
